src/prefect/engine.py

```diff
@@ -1852,19 +1852,24 @@ async def propose_state(
 
         link_state_to_result(state, result)
 
+    async def set_state_and_wait(set_state_func) -> OrchestrationResult:
```

Comment:
Maybe `set_state_and_handle_waits`? Can you add a docstring here that says we're doing this to prevent recursion depth

Reply:
Is it feasible to add a regression test that mocks sleep and has the server return a WAIT instruction 3000 times or w/e?

Reply:
!! I did not commit my test file. but that sounds a lot better than what I had, which was just checking it could handle a couple waits

Reply:
Updated and added tests that will WAIT recursion depth + buffer number of times